.test-infra/metrics/sync/github/sync_workflows.py

```diff
@@ -1,185 +1,203 @@
-        query = query + rowInsert
-        query += "),"
-    query = query[:-1] + ";"  
-    query = queryInsert + query
-    cursor.execute(query)
-    cursor.close()
-    connection.commit()
-    connection.close()
+      print('Failed to connect to DB; retrying in 1 minute')
+      print(e)
+      time.sleep(60)
+      i = i + 1
+      if i >= maxRetries:
+        print("Number of retries exceded ")
+        sys.exit(1)
+  return connection
+
+# Retries 9 times: 1s, 2s, 4s, 8s, 16s, ...
+@backoff.on_exception(backoff.expo, Exception, max_tries=9, factor=2)
+async def fetch(url, params=None, headers=None):
+  async with aiohttp.ClientSession() as session:
+    async with session.get(url, params=params, headers=headers) as response:
+      if response.status != 200:
+        raise aiohttp.ClientResponseError(
+          response.request_info,
+          response.history,
+          status=response.status,
+          message=response.reason,
+          headers=response.headers
+        )
+      return await response.json()
+
+async def fetch_workflow_data():
+  url = "https://api.github.com/repos/apache/beam/actions/workflows"
+  headers = {'Authorization': f'Bearer {GH_ACCESS_TOKEN}'}
+  page = 1
+  number_of_entries_per_page = 100 # The number of results per page (max 100)
+  query_options =\
+    {'branch': 'master', 'page': page, 'per_page': number_of_entries_per_page}
+
+  print("Start fetching recent workflow runs")
+  workflow_tasks = []
+  response = await fetch(url, query_options, headers)
+  while math.ceil(response['total_count'] / number_of_entries_per_page) >= page:
+    query_options = {
+      'branch': 'master',
+      'page': page,
+      'per_page': number_of_entries_per_page
+    }
+    task = asyncio.ensure_future(fetch(url, query_options, headers))
+    workflow_tasks.append(task)
+    page += 1
+
+  workflow_run_tasks = []
+  workflows_dict = {}
+  for completed_task in asyncio.as_completed(workflow_tasks):
+    response = await completed_task
+    workflows = response.get('workflows', [])
+    for workflow in workflows:
+      workflow_id = workflow['id']
+      workflow_name = workflow['name']
+      workflow_path = workflow['path']
+      result = re.search(r'(workflows\/.*)$', workflow_path)
+      if result:
+        workflow_path = result.group(1)
+
+      workflows_dict[workflow_id] =\
+        Workflow(workflow_id, workflow_name, workflow_path)
+
+      runs_url = f"{url}/{workflow_id}/runs"
+      page = 1
+      pages_to_fetch = math.ceil(
+        GH_NUMBER_OF_WORKFLOW_RUNS_TO_FETCH / number_of_entries_per_page
+      )
+      while pages_to_fetch >= page:
+        query_options = {
+          'branch': 'master',
+          'page': page,
+          'per_page': number_of_entries_per_page,
+          'exclude_pull_requests': 'true'
+        }
+        task = asyncio.ensure_future(fetch(runs_url, query_options, headers))
+        workflow_run_tasks.append(task)
```

Comment:
Can we skip cancelled runs here? I noticed that there are a bunch of failed runs in http://metrics.beam.apache.org/d/CTYdoxP4z/ga-post-commits-status?orgId=1 that are actually just cancelled, I think that's because we try unsuccessfully to handle those runs below. Alternately we can handle it below and skip them instead of just not setting their status.

Reply:
Added a filter to skip fictitious runs